src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+	trigger(id uint64, x string) error
+}
+
+type notifier struct {
+	mu       sync.RWMutex
+	requests map[uint64]chan string
+}
+
+func (ntf *notifier) register(id uint64) (<-chan string, error) {
+	ntf.mu.Lock()
+	defer ntf.mu.Unlock()
+	ch := ntf.requests[id]
+	if ch != nil {
+		return nil, fmt.Errorf("dup id %x", id)
+	}
+
+	ch = make(chan string, 1)
+	ntf.requests[id] = ch
+	return ch, nil
+}
+
+func (ntf *notifier) trigger(id uint64, x string) error {
+	ntf.mu.Lock()
+	ch, ok := ntf.requests[id]
+	if ch == nil || !ok {
+		ntf.mu.Unlock()
+		return fmt.Errorf("request ID %d not found", id)
+	}
+	delete(ntf.requests, id)
+	ntf.mu.Unlock()
+	ch <- x
+	close(ch)
+	return nil
+}
+```
+
+Alan now needs the equivalent to Go `sync.RWMutex`, and found multiple options:
+
+- [`std::sync::RwLock`](https://doc.rust-lang.org/std/sync/struct.RwLock.html)
+- [`parking_lot::RwLock`](https://docs.rs/parking_lot)
+
+Already losing confidence in Rust std, Alan instead chooses `parking_lot`, as it claims up to 5x faster performance than `std::sync::Mutex` (see [github](https://github.com/Amanieu/parking_lot#parking_lot)). After numeruous hours of trials and errors, Alan discovered that `parking_lot::RwLock` is not intended for async/future environments (see [github issue](https://github.com/Amanieu/parking_lot/issues/86)). Having to think about which library to use for thread and async programming, Alan appreciates the simplicity of Go concurrency where threads are effectively abstracted away from its users. Alan is now using [`async_std::sync::RwLock`](https://docs.rs/async-std/1.9.0/async_std/sync/struct.RwLock.html) which seems nicely integrated with Rust async programming.
+
+To send and receive events, Alan needs the equivalent of Go channel but is not sure about [`std::sync::mpsc::channel`](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html), as he sees two other options: [Flume](https://github.com/zesterer/flume) which claims to be much faster than std (see ["Flume, a 100% safe MPSC that's faster than std (from Reddit)"](https://www.reddit.com/r/rust/comments/fj17z6/flume_a_100_safe_mpsc_thats_faster_than_std_and/)), and [`crossbeam_channel`](https://docs.rs/crossbeam-channel/0.5.1/crossbeam_channel/). Having used crossbeam, Alan chose crossbeam channel:
+
+```rust
+use async_std::sync::RwLock;
+use crossbeam_channel::{self, unbounded};
+
+pub struct Notifier {
+    requests: RwLock<HashMap<u64, crossbeam_channel::Sender<String>>>,
+}
+
+impl Notifier {
+    pub fn new() -> Self {
+        Self {
+            requests: RwLock::new(HashMap::new()),
+        }
+    }
+
+    pub fn register(&self, id: u64) -> io::Result<crossbeam_channel::Receiver<String>> {
+        let mut _mu;
+        match self.requests.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+
+        let (request_tx, request_rx) = unbounded();
+        if _mu.get(&id).is_none() {
+            _mu.insert(id, request_tx);
```

Comment:
If this is used, why does it need to have underscore? Also, why is `_mu` which is a mutex guard named as `mutex`? Shouldn't it be `writer` or `requests`?